resolver/src/resolver_handle/mod.rs

```diff
@@ -59,167 +59,152 @@ impl DnsHandle for BasicResolverHandle {
+            fn(oneshot::Canceled) -> ResolveError
+        >,
+        G,
+        fn(F) -> G
+    >,
+    future::FutureResult<T, ResolveError>,
+>;
+
+/// Future returned by `LookupIp` requests to the background task.
+pub type BackgroundLookupIp = Background<LookupIp, LookupIpFuture, LookupIpFuture>;
+
+/// Future returned by lookup requests to the background task.
+pub type BackgroundLookup<T=Lookup, F=LookupFuture> = Background<T, LookupFuture, F>;
+
+/// Used by `ResolverHandle` for communicating with the background resolver task.
+enum Request {
+    /// Requests a lookup of the specified `RecordType`.
+    Lookup {
+        name: Name,
+        record_type: RecordType,
+        options: DnsRequestOptions,
+        tx: oneshot::Sender<LookupFuture>,
+    },
+    /// Requests an IP lookup for a name or IP address.
+    Ip {
+        maybe_name: ProtoResult<Name>,
+        maybe_ip: Option<RData>,
+        tx: oneshot::Sender<LookupIpFuture>,
+    },
 }
 
 macro_rules! lookup_fn {
-    ($p:ident, $f:ty, $r:path) => {
+    ($p:ident, $f:ty, $t:ty, $r:path) => {
 /// Performs a lookup for the associated type.
 ///
 /// *hint* queries that end with a '.' are fully qualified names and are cheaper lookups
 ///
 /// # Arguments
 ///
 /// * `query` - a string which parses to a domain name, failure to parse will return an error
-pub fn $p<N: IntoName>(&self, query: N) -> $f {
+pub fn $p<N: IntoName>(&self, query: N) -> BackgroundLookup<$t, $f> {
     let name = match query.into_name() {
         Ok(name) => name,
         Err(err) => {
-            return LookupFuture::error(self.client_cache.clone(), err).into();
+            return future::Either::B(future::err(err.into()));
         }
     };
 
-    self.inner_lookup(name, $r, DnsRequestOptions::default()).into()
+    self.inner_lookup(name, $r, DnsRequestOptions::default(),)
 }
     };
-    ($p:ident, $f:ty, $r:path, $t:ty) => {
+    ($p:ident, $f:ty, $t:ty, $r:path, $q:ty) => {
 /// Performs a lookup for the associated type.
 ///
 /// # Arguments
 ///
 /// * `query` - a type which can be converted to `Name` via `From`.
-pub fn $p(&self, query: $t) -> $f {
+pub fn $p(&self, query: $q) -> BackgroundLookup<$t, $f> {
     let name = Name::from(query);
-    self.inner_lookup(name, $r, DnsRequestOptions::default()).into()
+    self.inner_lookup(name, $r, DnsRequestOptions::default(),)
 }
     };
 }
 
-impl ResolverFuture {
-    /// Construct a new ResolverFuture with the associated Client and configuration.
+impl ResolverHandle {
+    /// Construct a new ResolverHandle with the associated Client and configuration.
     ///
     /// # Arguments
     ///
     /// * `config` - configuration, name_servers, etc. for the Resolver
     /// * `options` - basic lookup options for the resolver
+    ///
+    /// # Returns
+    ///
+    /// A tuple containing the new ResolverHandle and a future that drives the
+    /// background task that runs resolutions for the ResolverHandle.
     pub fn new(
```

Comment:
We should file an issue for a ResolverBuilder and then deprecate this function I think.

Reply:
In order to add additional configuration, or for some other reason?

Reply:
That was more a note for the project. For some reason renaming this type made me think that the new is potentially confusing.

Reply:
That's reasonable. Could also give the function a more descriptive name and deprecate `new`

Reply:
Given your comments in https://github.com/bluejekyll/trust-dns/pull/487#discussion_r190972819, I'm inclined to go with `AsyncResolver::new`, what do you think?